backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/codec/impl/LuceneBigDecimalFieldCodec.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.types.codec.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+import org.hibernate.search.backend.lucene.document.impl.LuceneDocumentBuilder;
+import org.hibernate.search.backend.lucene.logging.impl.Log;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.StoredField;
+import org.apache.lucene.index.IndexableField;
+
+public final class LuceneBigDecimalFieldCodec extends AbstractLuceneNumericFieldCodec<BigDecimal, Long> {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final int decimalScale;
+
+	public LuceneBigDecimalFieldCodec(boolean projectable, boolean sortable, BigDecimal indexNullAsValue, int decimalScale) {
+		super( projectable, sortable, indexNullAsValue );
+		this.decimalScale = decimalScale;
+	}
+
+	@Override
+	void doEncodeForProjection(LuceneDocumentBuilder documentBuilder, String absoluteFieldPath, BigDecimal value,
+			Long encodedValue) {
+		// we store the field for projections using its double representation
+		// we do not unscale it to a long using the decimalScale in this case
+		documentBuilder.addField( new StoredField( absoluteFieldPath, value.doubleValue() ) );
+	}
+
+	@Override
+	public BigDecimal decode(Document document, String absoluteFieldPath) {
+		IndexableField field = document.getField( absoluteFieldPath );
+
+		if ( field == null ) {
+			return null;
+		}
+
+		// decoding from stored value for projection
+		// so we can safety cast it to Double
+		Double doubleValue = (Double) field.numericValue();
+		BigDecimal bigDecimal = BigDecimal.valueOf( doubleValue );
+		return customStripTrailingZeros( bigDecimal );
+	}
+
+	@Override
+	public Long encode(BigDecimal value) {
+		return unscale( value );
+	}
+
+	@Override
+	public LuceneNumericDomain<Long> getDomain() {
+		return LuceneNumericDomain.LONG;
+	}
+
+	private Long unscale(BigDecimal value) {
+		BigDecimal rounded = value.setScale( decimalScale, RoundingMode.HALF_EVEN );
+		if ( value.scale() > decimalScale ) {
+			log.bigDecimalRounded( value, value.scale(), decimalScale );
+		}
+		return rounded.unscaledValue().longValue();
+	}
+
+	private static BigDecimal customStripTrailingZeros(BigDecimal bigDecimal) {
```

Comment:
Yes scale to 0 only if all decimal digits are zeros. I added, the method because Elasticsearch does the same, see `DecimalScaleIT#decimalScale_doesNotAffectProjections_exceedingZeros`.
If you want we can remove it, accepting different behaviors among backends. But I would like to keep it.